File: src/java/magma/compile/rule/FirstLocator.java
package magma.compile.rule;

import java.util.Optional;

public class FirstLocator implements Locator {
    @Override
    public Optional<Integer> locate(String input, String infix) {
        final var index = input.indexOf(infix);
        return index == -1 ? Optional.empty() : Optional.of(index);
    }
}

File: src/c/magma/app/rule/locate/FirstLocator.c
import magma.api.stream.Stream;import magma.api.stream.Streams;import java.util.Optional;struct FirstLocator(String infix){
	String unwrap(){
		return this.infix;
	}
	int length(){
		return this.infix.length();
	}
	Stream<Integer> locate(any* _ref_, String input){
		var index=input.indexOf(this.infix);
		return index==-1?Streams.empty():Streams.of(index);
	}
	Locator N/A(){
		return N/A.new();
	}
}